File: 10-28-20/gridTraveler.swift
// Swift doesn't like mutating arguments. Use the keyword `inout` to allow for this.
func gridTravelerHelper(_ m: Int, _ n: Int, _ memo: inout [String: Int]) -> Int {
  let key = "\(m), \(n)"
  if let val = memo[key] { // memo checking logic
    return val
  }

  if m == 0 || n == 0 {
    return 0
  }

  if m == 1 || n == 1 {
    return 1
  }

  let numWaysGoingDown = gridTravelerHelper(m - 1,n, &memo)
  let numWaysGoingRight = gridTravelerHelper(m, n - 1, &memo)
  memo[key] = numWaysGoingDown + numWaysGoingRight
  // The value at memo[key] *could* be nil so we use `!` to denote that we're sure there's a value.
  return memo[key]!
}

func gridTraveler(row: Int, column: Int) -> Int {
  // Helper function to pass in memo as a `var` (mutable variable)
  var memo = [String: Int]()
  return gridTravelerHelper(row, column, &memo)
}

print(gridTraveler(row: 2, column: 3)) // 3
print(gridTraveler(row: 3, column: 2)) // 3
print(gridTraveler(row: 3, column: 3)) // 6
print(gridTraveler(row: 16, column: 16)) // 155117520

File: 10-28-20/grid_traveler.rs
use std::collections::HashMap;

fn grid_traveler(m: u8, n: u8, memo: &mut HashMap<String, u128>) -> u128 {
  let key = format!("{}, {}", m, n).to_string();
  if memo.contains_key(&key) {
    return memo[&key];
  }

  if m == 0 || n == 0 {
    return 0;
  }

  if m == 1 || n == 1 {
    return 1;
  }

  let num_ways_going_down = grid_traveler(m - 1, n, memo);
  let num_ways_going_right = grid_traveler(m, n - 1, memo);
  memo.insert(key.clone(), num_ways_going_down + num_ways_going_right);
  memo[&key]
}

fn main() {
  println!("{}", grid_traveler(2, 3, &mut HashMap::new())); // 3
  println!("{}", grid_traveler(3, 2, &mut HashMap::new())); // 3
  println!("{}", grid_traveler(3, 3, &mut HashMap::new())); // 6
  println!("{}", grid_traveler(16, 16, &mut HashMap::new())); // 155117520
}